gogen/internal/gotypes/ordered_map.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// RoutingPolicy_PolicyDefinition_Statement represents an ordered-map element.
+type RoutingPolicy_PolicyDefinition_Statement struct {
+	DummyActions []string
+	Name         *string
+}
+
+// RoutingPolicy_PolicyDefinition_Statement_Map is a candidate ordered-map
+// implementation.
+type RoutingPolicy_PolicyDefinition_Statement_Map struct {
+	// TODO: Add a mutex here and add race tests after implementing
+	// ygot.Equal and evaluating the thread-safety of ygot.
+	//mu       sync.RWmutex
+	// keys contain the key order of the map.
+	keys []string
+	// valueMap contains the mapping from the statement key to each of the
+	// policy statements.
+	valueMap map[string]*RoutingPolicy_PolicyDefinition_Statement
+}
+
+// init initializes any uninitialized values.
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) init() {
+	if o.valueMap == nil {
+		o.valueMap = map[string]*RoutingPolicy_PolicyDefinition_Statement{}
+	}
+}
+
+// Keys returns a copy of the list's keys.
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) Keys() []string {
+	return append([]string{}, o.keys...)
+}
+
+// ValueSlice returns the current set of the list's values in order.
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) ValueSlice() []*RoutingPolicy_PolicyDefinition_Statement {
+	var values []*RoutingPolicy_PolicyDefinition_Statement
+	for _, key := range o.keys {
+		values = append(values, o.valueMap[key])
+	}
+	return values
+}
+
+// Len returns a size of RoutingPolicy_PolicyDefinition_Statement_Map
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) Len() int {
+	return len(o.keys)
+}
+
+// Get returns a value corresponding to the key. If the key is not found, the zero
+// value of K is returned with found being false.
+// Get is O(1).
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) Get(key string) *RoutingPolicy_PolicyDefinition_Statement {
+	val, _ := o.valueMap[key]
+	return val
+}
+
+// Delete deletes an element -- this is O(n) to keep the simple implementation.
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) Delete(key string) bool {
+	for i, k := range o.keys {
+		if k == key {
+			o.keys = append(o.keys[:i], o.keys[i+1:]...)
+			delete(o.valueMap, key)
+			return true
+		}
+	}
+	return false
+}
+
+// Append appends a policy statement, returning an error if the statement
+// already exists or if the key is unspecified.
+func (o *RoutingPolicy_PolicyDefinition_Statement_Map) Append(statement *RoutingPolicy_PolicyDefinition_Statement) error {
+	if statement == nil {
+		return errors.New("nil statement")
+	}
+	if statement.Name == nil {
+		return errors.New("nil key Name")
+	}
+	key := *statement.Name
+	if _, ok := o.valueMap[key]; ok {
+		return fmt.Errorf("duplicate key for list Statement %v", key)
+	}
+	o.keys = append(o.keys, key)
+	o.init()
+	o.valueMap[key] = statement
+	return nil
+}
+
+// Update updates a current policy statement, returning an error if the statement
+// doesn't exist or if the key is unspecified.
```

Comment:
i don't think Update and AppendNew would  be that useful, what use cases did you have in mind for them.

I'd rather have GetOrAppend(key) *RoutingPolicyDefinition

Reply:
I added `Update` as a basic dictionary operation. I don't have a specific use case in mind. I think it's ok to remove it if we want to keep to specific use cases.

`AppendNew` was added because Scott mentioned the current `New` will be useful alongside `Append`. This would be the equivalent for ordered maps.

As for `GetOrAppend`, this was in the draft I proposed originally since it follows existing conventions with unordered maps. I received feedback that this might be confusing because they want to know whether it was a Get or Append, since with ordered maps each element in general (e.g. BGP policies) cares about what other elements already exist, whereas in unordered maps each element often does not care about the existence of other elements (e.g. interfaces). So splitting it into `Get` and `AppendNew` helps the user write code that's less ambiguous. The stakeholders mentioned `New` and `Append` as the primary functionalities that they would use.

Reply:
I'll try to get a code review from the users.

Reply:
So I got feedback that `Append` is probably all they will need. For `AppendNew` I think given this API exists in the existing API method set, that it does not subtract value to have. I've removed `Update` though since I agree it might be a stretch for people to use it. PTAL

Reply:
here's my thoughts. 
1. GetOrCreate is by far the most commonly used function (at least in featureprofiles). Because it doesn't return an error, it is really convenient to use.  
2. Append: is very rarely used, though it's a still valuable and should exist
3. AppendNew: I don't see this in the existing API set, and personally don't see when i'd use it.

i'm fine with the API as is.

Reply:
I agree with 1&2 from a current-API perspective, the current PR reflects iteration after feedback. Ordered-maps are different than unordered-maps and IMO this method set has greater usability than simply copying what the existing unordered-map API does.

For 3 AppendNew corresponds to New for unordered maps.